src/main/java/apoc/util/google/cloud/GCStorageURLConnection.java

```diff
@@ -0,0 +1,95 @@
+package apoc.util.google.cloud;
+
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import org.apache.commons.lang.StringUtils;
+import org.neo4j.util.VisibleForTesting;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class GCStorageURLConnection extends URLConnection {
+
+    enum AuthType { NONE, SERVICE }
+
+    private Blob blob;
+    public GCStorageURLConnection(URL url) {
+        super(url);
+    }
+
+    @VisibleForTesting
+    public Storage getStorage(URI uri) {
+        Storage storage;
+        Map<String, String> queryParams = getQueryParams(uri);
+        AuthType authenticationType = AuthType.valueOf(queryParams.getOrDefault("authenticationType", AuthType.NONE.toString()));
+        switch (authenticationType) {
+            case SERVICE:
+                String googleAppCredentialsEnv = System.getenv("GOOGLE_APPLICATION_CREDENTIALS");
```

Comment:
can't it be passed in other means? e.g. in server setups those might not be settable?

Reply:
As the first implementation, I tried to stick to the best practice, but I added the authentication type because as you underscored there can be someone that uses an alternative way to authenticate to the service, so we can improve that in a community-driven way. WDYT?